### PR TITLE
Omptarget Updates

### PR DESCRIPTION
**Pull request by crtrott**
Significant Update to OpenMPTarget

- requires C++17
- works on Intel GPUs
- Compilation and tests fixed/worked around so that full build/test passes.
- Fixed some stuff for reference counting.

### CODE REVIEW

File: core/src/impl/Kokkos_SharedAlloc.hpp [hunk 1]
```diff
@@ -48,6 +48,12 @@
 #include <cstdint>
 #include <string>
 
+#if defined(KOKKOS_ENABLE_OPENMPTARGET) && defined(KOKKOS_ENABLE_CXX17)
+#define KOKKOS_IMPL_IF_ON_HOST if constexpr (omp_is_initial_device() == true)
+#else
+#define KOKKOS_IMPL_IF_ON_HOST if (true)
+#endif
```
Comment (dalg24) [inline core/src/impl/Kokkos_SharedAlloc.hpp:55]: #undef it at the end of the file and comment you do so

File: containers/src/Kokkos_ScatterView.hpp [hunk 2]
```diff
@@ -199,35 +211,30 @@ struct ScatterValue;
 
 template <typename ValueType, typename DeviceType>
 struct ScatterValue<ValueType, Kokkos::Experimental::ScatterSum, DeviceType,
-                    Kokkos::Experimental::ScatterNonAtomic>
-    : Sum<ValueType, DeviceType> {
+                    Kokkos::Experimental::ScatterNonAtomic> {
+  ValueType& value;
```
Comment (dhollman) [inline containers/src/Kokkos_ScatterView.hpp:215]: Did you mean for this to be public? If so, putting `public:` two lines below makes it look like you didn't intend to do that.

File: containers/src/Kokkos_ScatterView.hpp [hunk 3]
```diff
@@ -237,29 +244,26 @@ struct ScatterValue<ValueType, Kokkos::Experimental::ScatterSum, DeviceType,
  update(rhs) and reset() functions. */
 template <typename ValueType, typename DeviceType>
 struct ScatterValue<ValueType, Kokkos::Experimental::ScatterSum, DeviceType,
-                    Kokkos::Experimental::ScatterAtomic>
-    : Sum<ValueType, DeviceType> {
+                    Kokkos::Experimental::ScatterAtomic> {
+  ValueType& value;
```
Comment (dhollman) [inline containers/src/Kokkos_ScatterView.hpp:248]: Same here

File: containers/src/Kokkos_ScatterView.hpp [hunk 4]
```diff
@@ -199,35 +211,30 @@ struct ScatterValue;
...
-    this->join(this->reference(), ValueType(-rhs));
-  }
-  KOKKOS_FORCEINLINE_FUNCTION void operator--() {
-    this->join(this->reference(), ValueType(-1));
-  }
-  KOKKOS_FORCEINLINE_FUNCTION void operator--(int) {
-    this->join(this->reference(), ValueType(-1));
+    update(ValueType(-rhs));
   }
+  KOKKOS_FORCEINLINE_FUNCTION void operator--() { update(ValueType(-1)); }
```
Comment (dhollman) [inline containers/src/Kokkos_ScatterView.hpp:230]: You probably should have a reason for not using `{}` initialization with numeric literals to avoid accidental narrowing.

File: containers/src/Kokkos_ScatterView.hpp [hunk 5]
```diff
@@ -199,35 +211,30 @@ struct ScatterValue;
...
+  KOKKOS_FORCEINLINE_FUNCTION void operator--(int) { update(ValueType(-1)); }
```
Comment (dhollman) [inline containers/src/Kokkos_ScatterView.hpp:231]: Why do these return `void`?

File: core/src/impl/Kokkos_Atomic_Generic_Secondary.hpp [hunk 6]
```diff
@@ -72,5 +72,15 @@ KOKKOS_INLINE_FUNCTION void atomic_sub(volatile T* const dest, const T val) {
...
+template <typename T>
+KOKKOS_INLINE_FUNCTION void atomic_mul(volatile T* const dest, const T val) {
+  (void)atomic_fetch_mul(dest, val);
+}
+
+template <typename T>
+KOKKOS_INLINE_FUNCTION void atomic_div(volatile T* const dest, const T val) {
+  (void)atomic_fetch_div(dest, val);
+}
+
```
Comment (dhollman) [inline core/src/impl/Kokkos_Atomic_Generic_Secondary.hpp:84]: What does this have to do with OMPTarget?

File: core/src/impl/Kokkos_SharedAlloc.hpp [hunk 7]
```diff
@@ -48,6 +48,12 @@
 #include <cstdint>
 #include <string>
 
+#if defined(KOKKOS_ENABLE_OPENMPTARGET) && defined(KOKKOS_ENABLE_CXX17)
+#define KOKKOS_IMPL_IF_ON_HOST if constexpr (omp_is_initial_device() == true)
+#else
+#define KOKKOS_IMPL_IF_ON_HOST if (true)
```
Comment (dhollman) [inline core/src/impl/Kokkos_SharedAlloc.hpp:54]: Comment explaining why this is sufficient on pre-C++17, maybe?

File: core/unit_test/TestReduce.hpp [hunk 8]
```diff
@@ -510,12 +512,13 @@ TEST(TEST_CATEGORY, int_combined_reduce_mixed) {
 
   auto result3_v = Kokkos::View<int64_t, Kokkos::HostSpace>{"result3_v"};
 
-  Kokkos::parallel_reduce(nw, functor_type(nw), result1_v, result2,
+  Kokkos::parallel_reduce("int_combined-reduce_mixed",
+                          Kokkos::RangePolicy<TEST_EXECSPACE>(0, nw),
+                          functor_type(nw), result1_v, result2,
                           Kokkos::Sum<int64_t, Kokkos::HostSpace>{result3_v});
```
Comment (dhollman) [inline core/unit_test/TestReduce.hpp:518]: Is this a driveby change on combined reducers? If so, should it be in a separate pull request?

File: containers/src/Kokkos_ScatterView.hpp (hunk 4 above)
Comment (mhoemmen) [reply containers/src/Kokkos_ScatterView.hpp:230]: Wait, what if `ValueType` is unsigned?  It should still be OK, right?

Comment (dalg24) [reply containers/src/Kokkos_ScatterView.hpp:230]: > Wait, what if `ValueType` is unsigned? It should still be OK, right?

It will do the right thing

File: cmake/kokkos_enable_devices.cmake [hunk 9]
```diff
@@ -75,6 +78,7 @@ SET(ClangOpenMPFlag -fopenmp=libomp)
 
   COMPILER_SPECIFIC_FLAGS(
     Clang      ${ClangOpenMPFlag} -Wno-openmp-mapping
+    IntelClang -fiopenmp -Wno-openmp-mapping
```
[the remark this answers is not in the document]
Comment (crtrott) [reply cmake/kokkos_enable_devices.cmake:81]: its not fopenmp=libomp, its fiopenmp with an "i", because its not really clang ….

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp [hunk 10]
```diff
@@ -50,6 +50,8 @@
 #include <OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp>
 #include <impl/Kokkos_FunctorAdapter.hpp>
 
+#define KOKKOS_IMPL_MDRANGE_USE_NO_TILES
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp:53]: it is a workaround

File: core/src/impl/Kokkos_ClockTic.hpp [hunk 11]
```diff
@@ -77,9 +80,11 @@ uint64_t clock_tic(void) noexcept {
   // Get clock register
   return hc::__clock_u64();
 #elif defined(KOKKOS_ENABLE_OPENMPTARGET)
+  /*
   return (uint64_t)std::chrono::high_resolution_clock::now()
       .time_since_epoch()
-      .count();
+      .count();*/
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/src/impl/Kokkos_ClockTic.hpp:86]: because it did work on proper clang but not on intel clang, and we don't know why

File: core/unit_test/TestReduce.hpp (hunk 8 above)
Comment (crtrott) [reply core/unit_test/TestReduce.hpp:518]: its not necessarily a drive-by change. It is a change specifically necessary for OpenMPTarget since OpenMPTarget doesn't have something like UVM or HostPinned memory which was used previously for these tests.

File: core/src/impl/Kokkos_SharedAlloc.hpp (hunk 7 above)
Comment (crtrott) [reply core/src/impl/Kokkos_SharedAlloc.hpp:54]: it is not, but OpenMPTarget now requires C++17, I can remove the KOKKOS_ENABLE_CXX17 here.

File: core/src/impl/Kokkos_Atomic_Generic_Secondary.hpp (hunk 6 above)
Comment (crtrott) [reply core/src/impl/Kokkos_Atomic_Generic_Secondary.hpp:84]: nothing which is why its also in this PR: #3162

File: containers/src/Kokkos_ScatterView.hpp (hunk 5 above)
Comment (crtrott) [reply containers/src/Kokkos_ScatterView.hpp:231]: they did before. And yeah we probably should fix that as general ScatterView overhaul. Also #3162

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp:53]: You must add a comment

File: core/src/impl/Kokkos_ClockTic.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/impl/Kokkos_ClockTic.hpp:86]: Comment or remove

**Comment by masterleinad** [comment]
The formatting is off and the `CUDA-11.0-NVCC-C++17-RDC` CI complains:
```
/var/jenkins/workspace/Kokkos/install/include/Cuda/Kokkos_Cuda_Parallel.hpp(471): error: calling a constexpr __host__ function("operator()") from a __device__ function("exec_range") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.
          detected during:
            instantiation of "std::enable_if<std::is_same<TagType, void>::value, void>::type Kokkos::Impl::ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::Cuda>::exec_range<TagType>(Kokkos::Impl::ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::Cuda>::Member) const [with FunctorType=lambda [](int)->void, Traits=<Kokkos::CudaUVMSpace::execution_space>, TagType=void]" 
(493): here
            instantiation of "void Kokkos::Impl::ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::Cuda>::operator()() const [with FunctorType=lambda [](int)->void, Traits=<Kokkos::CudaUVMSpace::execution_space>]" 
/var/jenkins/workspace/Kokkos/install/include/Cuda/Kokkos_Cuda_KernelLaunch.hpp(124): here
            instantiation of "void Kokkos::Impl::cuda_parallel_launch_local_memory(DriverType) [with DriverType=Kokkos::Impl::ParallelFor<lambda [](int)->void, Kokkos::RangePolicy<Kokkos::CudaUVMSpace::execution_space>, Kokkos::CudaUVMSpace::execution_space>]" 
/var/jenkins/workspace/Kokkos/install/include/Cuda/Kokkos_Cuda_KernelLaunch.hpp(478): here
            instantiation of "cudaFuncAttributes Kokkos::Impl::CudaParallelLaunch<DriverType, Kokkos::LaunchBounds<0U, 0U>, Kokkos::Impl::Experimental::CudaLaunchMechanism::LocalMemory>::get_cuda_func_attributes() [with DriverType=Kokkos::Impl::ParallelFor<lambda [](int)->void, Kokkos::RangePolicy<Kokkos::CudaUVMSpace::execution_space>, Kokkos::CudaUVMSpace::execution_space>]" 
(501): here
            instantiation of "void Kokkos::Impl::ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::Cuda>::execute() const [with FunctorType=lambda [](int)->void, Traits=<Kokkos::CudaUVMSpace::execution_space>]" 
/var/jenkins/workspace/Kokkos/install/include/Kokkos_Parallel.hpp(176): here
            instantiation of "void Kokkos::parallel_for(const ExecPolicy &, const FunctorType &, const std::__cxx11::string &, std::enable_if<Kokkos::is_execution_policy<ExecPolicy>::value, void>::type *) [with ExecPolicy=Kokkos::RangePolicy<Kokkos::CudaUVMSpace::execution_space>, FunctorType=lambda [](int)->void]" 
/var/jenkins/workspace/Kokkos/install/include/Kokkos_Parallel.hpp(218): here
            instantiation of "void Kokkos::parallel_for(const std::__cxx11::string &, const ExecPolicy &, const FunctorType &) [with ExecPolicy=Kokkos::RangePolicy<Kokkos::CudaUVMSpace::execution_space>, FunctorType=lambda [](int)->void]" 
/var/jenkins/workspace/Kokkos/core/unit_test/TestViewSubview.hpp(973): here
            instantiation of "void TestViewSubview::Impl::test_1d_assign_impl<Space,LayoutSub,Layout,LayoutOrg,MemTraits>() [with Space=Kokkos::CudaUVMSpace, LayoutSub=Kokkos::LayoutLeft, Layout=Kokkos::LayoutLeft, LayoutOrg=Kokkos::LayoutLeft, MemTraits=void]" 
/var/jenkins/workspace/Kokkos/core/unit_test/TestViewSubview.hpp(1734): here
            instantiation of "void TestViewSubview::test_1d_assign<Space,MemTraits>() [with Space=Kokkos::CudaUVMSpace, MemTraits=void]" 
/var/jenkins/workspace/Kokkos/core/unit_test/cuda/TestCuda_SubView_c01.cpp(51): here
```

File: core/perf_test/CMakeLists.txt [hunk 12]
```diff
@@ -53,7 +53,14 @@ IF(Kokkos_ENABLE_HIP)
...
     PerfTest_ExecSpacePartitioning.cpp
-    )
+  )
+ENDIF()
+
+IF(Kokkos_ENABLE_OPENMPTARGET)
+  LIST(REMOVE_ITEM SOURCES
+    PerfTest_CustomReduction.cpp
+    PerfTest_ExecSpacePartitioning.cpp
+  )
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/perf_test/CMakeLists.txt:64]: done

File: core/perf_test/CMakeLists.txt [hunk 13]
```diff
@@ -87,8 +94,10 @@ KOKKOS_ADD_EXECUTABLE_AND_TEST(
   CATEGORIES PERFORMANCE
 )
 
-KOKKOS_ADD_EXECUTABLE_AND_TEST(
-  PerformanceTest_TaskDag
-  SOURCES test_taskdag.cpp
-  CATEGORIES PERFORMANCE
-)
+IF(NOT Kokkos_ENABLE_OPENMPTARGET)
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/perf_test/CMakeLists.txt:98]: done

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 14]
```diff
@@ -61,6 +61,9 @@ template <class Reducer>
 struct OpenMPTargetReducerWrapper {
   typedef typename Reducer::value_type value_type;
 
+// This pragma omp declare target should not be necessary, but Intel compiler
+// fails without it
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:66]: done

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (crtrott) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_MDRange.hpp]: done

File: core/src/impl/Kokkos_ClockTic.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (crtrott) [reply core/src/impl/Kokkos_ClockTic.hpp]: removed

File: core/src/impl/Kokkos_SharedAlloc.hpp [hunk 15]
```diff
@@ -48,6 +48,12 @@
 #include <cstdint>
 #include <string>
 
+#if defined(KOKKOS_ENABLE_OPENMPTARGET) && defined(KOKKOS_ENABLE_CXX17)
+#define KOKKOS_IMPL_IF_ON_HOST if constexpr (omp_is_initial_device() == true)
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/src/impl/Kokkos_SharedAlloc.hpp]: done

File: core/src/impl/Kokkos_SharedAlloc.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/impl/Kokkos_SharedAlloc.hpp:57]: done

**Comment by crtrott** [comment]
I ignored the ScatterView comments since that came in via a different PR, and we agreed to do another overhaul on that.

File: cmake/kokkos_enable_devices.cmake (hunk 9 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply cmake/kokkos_enable_devices.cmake:81]: Right but you define a variable line 48 and never use it

File: core/perf_test/CMakeLists.txt [hunk 16]
```diff
@@ -49,11 +49,19 @@ SET(SOURCES
   )
 
 IF(Kokkos_ENABLE_HIP)
-# FIXME requires TeamPolicy
+# FIXME HIP requires TeamPolicy
```
Comment (dalg24) [inline core/perf_test/CMakeLists.txt:52]: We use the FIXME_HIP token

File: core/perf_test/CMakeLists.txt [hunk 17]
```diff
@@ -49,11 +49,19 @@ SET(SOURCES
...
   LIST(REMOVE_ITEM SOURCES
     PerfTest_CustomReduction.cpp
     PerfTest_ExecSpacePartitioning.cpp
-    )
+  )
+ENDIF()
+
+IF(Kokkos_ENABLE_OPENMPTARGET)
+# FIXME OPENMPTARGET requires TeamPolicy Reductions and Custom Reduction
```
Comment (dalg24) [inline core/perf_test/CMakeLists.txt:60]: You are mixing `FIXME OPENMPTARGET` with `WORKAROUND OPENMPTARGET`.  The point was to have unique token that can be searched for.

File: core/perf_test/CMakeLists.txt (hunk 16 above)
Comment (crtrott) [reply core/perf_test/CMakeLists.txt:52]: Apparently you don't because it only said FIXME before ;-)

File: core/perf_test/CMakeLists.txt (hunk 17 above)
Comment (crtrott) [reply core/perf_test/CMakeLists.txt:60]: I never used WORKAROUND in a cmake file, so I followed what was in the file.

File: cmake/kokkos_enable_devices.cmake (hunk 9 above)
[the remark this answers is not in the document]
Comment (crtrott) [reply cmake/kokkos_enable_devices.cmake:81]: probably because it wasn't used there before some recent update, which came into this branch via a rebase
